chain/jsonrpc-primitives/src/types/query.rs

```diff
@@ -0,0 +1,151 @@
+use serde::{Deserialize, Serialize};
+use serde_json::Value;
+
+#[derive(Serialize, Deserialize)]
+pub struct RpcQueryRequest {
+    #[serde(flatten)]
+    pub block_reference: near_primitives::types::BlockReference,
+    #[serde(flatten)]
+    pub request: near_primitives::views::QueryRequest,
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum RpcQueryError {
+    #[error("IO Error: {0}")]
+    IOError(String),
+    #[error("There are no fully synchronized blocks yet")]
+    NotSyncedYet,
+    #[error("The node does not track the shard")]
+    DoesNotTrackShard,
+    #[error("Invalid account ID {0}")]
+    InvalidAccount(near_primitives::types::AccountId),
+    #[error("Account ID {0} has never been observed on the node")]
+    AccountDoesNotExist(near_primitives::types::AccountId),
+    #[error("Contract code for contract ID {0} has never been observed on the node")]
+    ContractCodeDoesNotExist(near_primitives::types::AccountId),
+    #[error("Access key for public key {0} has never been observed on the node")]
+    AccessKeyDoesNotExist(String),
+    #[error("VM error occurred: {0}")]
+    VMError(String),
+    #[error("The node reached its limits. Try again later. More details: {0}")]
+    InternalError(String),
+    // NOTE: Currently, the underlying errors are too broad, and while we tried to handle
+    // expected cases, we cannot statically guarantee that no other errors will be returned
+    // in the future.
+    // TODO #3851: Remove this variant once we can exhaustively match all the underlying errors
+    #[error("It is a bug if you receive this error type, please, report this incident: https://github.com/near/nearcore/issues/new/choose. Details: {0}")]
+    Unreachable(String),
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct RpcQueryResponse {
+    #[serde(flatten)]
+    pub query_response: near_primitives::views::QueryResponse,
+}
+
+impl RpcQueryRequest {
+    pub fn parse(value: Option<Value>) -> Result<RpcQueryRequest, crate::errors::RpcParseError> {
+        let query_request = if let Ok((path, data)) =
+            crate::utils::parse_params::<(String, String)>(value.clone())
+        {
+            let data = near_primitives_core::serialize::from_base(&data)
```

Comment:
Please, preserve the comment we had in the original `jsonrpc/lib.rs` implementation about soft-deprecation